File: src/test/java/org/gecko/model/ElementTest.java
package org.gecko.model;

import org.gecko.exceptions.MissingViewModelElementException;
import org.gecko.exceptions.ModelException;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class ElementTest {
    static Element element;

    @BeforeAll
    static void setUp() {
        assertThrows(ModelException.class, () -> element = new Element(-1) {
            @Override
            public void accept(ElementVisitor visitor) throws ModelException, MissingViewModelElementException {

            }
        });
        assertDoesNotThrow(() -> element = new Element(0) {
            @Override
            public void accept(ElementVisitor visitor) throws ModelException, MissingViewModelElementException {

            }
        });
    }

    @Test
    void testHashCode() {
        assertNotEquals(0, element.hashCode());
    }

    @Test
    void testEquals() {
        assertEquals(element, element);

        final Element[] other = new Element[4];
        assertDoesNotThrow(() -> other[0] = new Element(1) {
            @Override
            public void accept(ElementVisitor visitor) throws ModelException, MissingViewModelElementException {

            }
        });
        assertDoesNotThrow(() -> other[1] = new Element(0) {
            @Override
            public void accept(ElementVisitor visitor) throws ModelException, MissingViewModelElementException {

            }
        });

        assertDoesNotThrow(() -> other[2] = new State(2, "state"));
        assertDoesNotThrow(() -> other[3] = new State(0, "state"));

        assertNotEquals(element, other[0]);
        assertEquals(element, other[1]);
        assertNotEquals(element, other[2]);
        assertEquals(element, other[3]);
        assertNotEquals(null, element);
    }
}

File: src/test/kotlin/org/gecko/model/ElementTest.kt
package org.gecko.model

import org.gecko.exceptions.MissingViewModelElementException
import org.gecko.exceptions.ModelException
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.ThrowingSupplier

class ElementTest {
    @Test
    fun testHashCode() {
        Assertions.assertNotEquals(0, element.hashCode())
    }

    @Test
    fun testEquals() {
        Assertions.assertEquals(element, element)

        val other = arrayOfNulls<Element>(4)
        Assertions.assertDoesNotThrow(ThrowingSupplier<Element> {
            other[0] = object : Element(1) {
                @Throws(ModelException::class, MissingViewModelElementException::class)
                override fun accept(visitor: ElementVisitor) {
                }
            }
        })
        Assertions.assertDoesNotThrow(ThrowingSupplier<Element> {
            other[1] = object : Element(0) {
                @Throws(ModelException::class, MissingViewModelElementException::class)
                override fun accept(visitor: ElementVisitor) {
                }
            }
        })

        Assertions.assertDoesNotThrow(ThrowingSupplier<Element> { other[2] = State(2, "state") })
        Assertions.assertDoesNotThrow(ThrowingSupplier<Element> { other[3] = State(0, "state") })

        Assertions.assertNotEquals(element, other[0])
        Assertions.assertEquals(element, other[1])
        Assertions.assertNotEquals(element, other[2])
        Assertions.assertEquals(element, other[3])
        Assertions.assertNotEquals(null, element)
    }

    companion object {
        var element: Element? = null

        @BeforeAll
        fun setUp() {
            Assertions.assertThrows(ModelException::class.java) {
                element = object : Element(-1) {
                    @Throws(ModelException::class, MissingViewModelElementException::class)
                    override fun accept(visitor: ElementVisitor) {
                    }
                }
            }
            Assertions.assertDoesNotThrow(ThrowingSupplier<Element> {
                element = object : Element(0) {
                    @Throws(ModelException::class, MissingViewModelElementException::class)
                    override fun accept(visitor: ElementVisitor) {
                    }
                }
            })
        }
    }
}
